### ISSUE TITLE
Can't replicate the stated performance improvements with cross-modal training

### ISSUE THREAD

**Issue by robinnarsinghranabhat** [linzhiqiu/cross_modal_adaptation#21]
## Approach : 
I wanted to utilize "Cross-modal training",  to use additional text labels as training examples. But it doesn't give better accuracy than 
"unimodal-training". I followed exact steps and verified things in the codebase as well. 

I can't get claimed results least for  : 
`dtd`, `oxford-pets` & `oxford-flowers` for `8-Shot` and `16-Shot` adaptation , and at different seed values.

I ran tests both with and without the additional Augmentation with "flip". 

Example Procedure for `dtd` : 

1. Prepare Features with "none" and "flip"
```
python features.py --dataset dtd --train-shot 16 --seed 1 --clip-encoder RN50 --image-layer-idx 0 --text-augmentation hand_crafted --image-augmentation flip --image-views 1

python features.py --dataset dtd --train-shot 16 --seed 1 --clip-encoder RN50 --image-layer-idx 0 --text-augmentation hand_crafted --image-augmentation flip --image-views 1
```

2. Unimodal Training : 
```
python train.py --modality uni_modal --classifier_head linear --classifier_init zeroshot --logit 4.60517 --hyperparams linear --dataset dtd --train-shot 16 --seed 1 --clip-encoder RN50 --image-layer-idx 0 --text-augmentation hand_crafted --image-augmentation flip --image-views 1
```

3. Crossmodal Training : 
```
python train.py --modality cross_modal --classifier_head linear --classifier_init zeroshot --logit 4.60517 --hyperparams linear --dataset dtd --train-shot 16 --seed 1 --clip-encoder RN50 --image-layer-idx 0 --text-augmentation hand_crafted --image-augmentation flip --image-views 1
```

Results : 
Best Unimodal Test-Set Accuracy : 
optim_adamw-lr_0.0001-wd_0.0-bs_8-iters_12800 7/12
`{'val_acc': 0.6808510638297872, 'iter': 800, 'test_accs': {'head': 0.6796690307328606, 'head_wiseft_0.5': 0.625886524822695}}`

Best Crossmodal Test-Set Accuracy : 
optim_adamw-lr_0.0001-wd_0.0001-bs_8-iters_12800 11/12
`{'val_acc': 0.6808510638297872, 'iter': 1100, 'test_accs': {'head': 0.6713947990543735, 'head_wiseft_0.5': 0.6128841607565012}}`

As you can see, unimodal training performance is infact slightly better.

**Comment by linzhiqiu** [comment]
Thanks for your interest in our work. Have you compared your results with Table 10 in our paper? From a quick glance, it looks like the numbers you're getting for cross-modal linear probing are quite similar to ours.

I’d like to emphasize that few-shot experiments tend to have high variance, so if possible, try running at least three random seeds to compute the mean and standard deviation, as reported in Table 10. For instance, you can see that the 16-shot DTD experiment had a standard deviation of 1.5%. This will help ensure your results are statistically meaningful.

Additionally, keep in mind that PyTorch/CUDA versions can introduce numerical differences depending on the machine setup.

In general, unimodal and cross-modal linear probing should converge to similar results as the number of labeled images increases. If you’re working with more shots, you might also consider partially fine-tuning the visual encoder to better adapt to the new distribution. Some of my co-authors have done follow-up work showing that fine-tuning the vision encoder with additional images can significantly improve performance: https://arxiv.org/abs/2406.11148

Hope this helps! Let me know if you have any other questions.

**Comment by robinnarsinghranabhat** [comment]
Appreciate this quick, detailed response. 

```... few-shot experiments tend to have high variance```

To maintain statistical validity, I tested extensively on `Oxford-Flowers`, `Oxford-Pets` and `DTD` ( For [1,2,3] seeds, and for [1,2,4,8,16] shots. In all cases, `the results were inclined towards SIMPLE UNIMODAL FINETUNING STILL BEING BETTER`. 


```it looks like the numbers you're getting for cross-modal linear probing are quite similar to ours```

Yes, results are similar. But the problem still with, unimodal is performing better than cross-modal, at least on above datasets.


> that PyTorch/CUDA versions can introduce numerical differences
Respectfully, I find it incredibly hard to believe this as the case. As, the results I am getting are consistent.


I am not here to scrutinize your work. In fact, I find this idea really interesting and wanted to check if training with cross-modal information from another domain could help with Noise-robustness.  

But now I am even more confused as I can't replicate findings. I am scared if the other past finetuning works (Coop, TIA-Adapter, Clip-adapter) only worked better, because of careful massaging of parameters, without general significance of the work itself. 

All I want is, my wishful thinking, for cross-modal to be better than unimodal in a simplified apples-to-apples comparison scenario :)  
Any help is greatly appreciated !

**Comment by linzhiqiu** [comment]
Hey,  

Of course, and thank you for your deep interest in our work! Since it’s been two years since I last touched this codebase, even I’m a bit rusty on the argparse arguments I originally defined.  

Looking at your script again (thanks for providing it), it seems that your unimodal linear probing setup **isn’t truly unimodal** because you initialized the linear classifier weights using the text features by specifying `--classifier_init zeroshot` instead of `--classifier_init random`. Here, *zeroshot* refers to **Zero-Shot CLIP** (as the original CLIP paper), which literally uses text features as linear classifier weights.  

This was my mistake, and I’m honestly surprised no one caught this bug in the past two years (I guess no one wants to run vanilla linear probing again): https://github.com/linzhiqiu/cross_modal_adaptation/blob/7641c5a3593906984b7c960276774a81923d3434/linear_probe.sh#L90

That line should be changed from "zeroshot" to "random"! Thanks for pointing it out! I’ll update the bash script soon (honestly, using a bash script for this wasn’t the best idea—it’s too hard to read), but I want to make sure you’re fully clear on this issue first.

We actually conducted an ablation in **Table 12** (*Ablation results for text-based vs. random initialization for linear classifier weight*). While I don’t have per-dataset results (which can be run using this codebase), you may check out the **average performance across all datasets** in Table 12. From that, you can see:  

1. **Using text features to initialize linear probing can make a huge difference in both unimodal and cross-modal settings**, especially in **1/2/4-shot settings**.  
2. **When averaging across all datasets, cross-linear probing still outperforms unimodal linear probing initialized with text features.**  

![Image](https://github.com/user-attachments/assets/57b94bdf-76c4-4510-8bb6-ba7d246dd6c6)

Regarding your point about **CoOp/Tip-Adapter/CLIP-Adapter**—to be honest, I’m not a big fan of these approaches, as they mostly borrow fancy concepts from NLP and apply them to CV. However, the **biggest performance gain** (compared to purely unimodal linear probing with randomly initialized weights) comes from leveraging **multimodal (language) information**.  Our paper titled *"Multimodality Helps Unimodality"* just shows there is a simpler way than CoOp/Tip-Adapter to utilize the language modality to achieve the same or even higher performance boost.

P.S. Again, I could be missing something since I haven’t worked on this in a while. Feel free to shoot me any questions if you have more, and I really appreciate you pointing this out!

**Comment by robinnarsinghranabhat** [comment]
Thanks ! that's why I love open source science. Polishing each other's work. 

With randomly initialized linear layer, I do see effect of cross-modal information :) 
Unfortunately, I won't be able to test this for imagenet dataset.

Also, I created a PR with mps support for mac users :). The branch could just be pointed out in just readme as well. Cuz PR reviewing is booring right :)  


If possible, do you think I could ask you further questions, I come across, over some platform ?  
I promise not to bug you unless necessary !

**Comment by linzhiqiu** [comment]
Thank you for the PR! I wasn’t aware of the MPS device before, but I’m open to this change for added flexibility. I’ve made a few stylistic commits and have a small question regarding the PR.

Feel free to reach out via my personal email (zl279@cornell.edu) or school email (zhiqiul@andrew.cmu.edu) if you'd like to discuss anything privately. Otherwise, GitHub issues/PRs work perfectly.
